### ISSUE TITLE
Remaps buttons gpio to virtual USB Keyboard

### ISSUE THREAD

**Issue by Laddyacris** [digitalLumberjack/recalbox-os#285]
ADAFRUIT RETROGAME UTILITY: remaps buttons on Raspberry Pi GPIO header
to virtual USB keyboard presses. 
https://github.com/adafruit/Adafruit-Retrogame
ça permet de créer un clavier usb virtuel sur les gpio , ça pourrait être utile pour les emulateurs ayant recourt au clavier pour jouer.

**Comment by ian57** [comment]
This is already done by the kernel module mk_arcade_joystick_rpi : The Raspberry Pi GPIO Joystick Driver ;)

**Comment by Laddyacris** [comment]
ah oui ?
mon idée etait de mettre un bouton reset qui permet de sortir d emulationstation, comment faire comprendre au gpio via mk_arcade de le faire ?

**Comment by ian57** [comment]
To exit from an emu, or from ES? I'm very new to RecalBox, but I think you can exit emu with a combo like StartP1 + StartP2. This config has to be done for libretroarch. But be careful, in retroarch, the select button cannot be used as the input_enable_hotkey_bt, it is ignored.

For ES, without the --no-exit option, ES show the "Quit ES" item in the main menu to go back to console.

If you want to use retrogame, take look at my github repo https://github.com/ian57/Raspicade-Retrogame-2Player-Pi2 of retrogame

**Comment by Laddyacris** [comment]
Je pense que tu n'as pas compris mon objectif "idée"
Via joysktick la combinaison de touche hotkey+start pour sortir d'un emulateur fonctionne.
le but est de détourner un pin gpio pour lui attribuer la fonction "escape" d'un clavier pour sortir d'un emulateur et non du fontend.

**Comment by ian57** [comment]
Yes I understood, but why add a button to do escape... Combo is better no?

To add a button on a pin, you'll need to modify mk_arcade_joystick_rpi source code... or we'll have to do it?

As far as i know Retrogame is not included in Recalbox.

**Comment by Laddyacris** [comment]
Pour aider violene sur le forum avec le bouton reset de sa vieille console snes et de son retrode.

**Comment by digitalLumberjack** [comment]
The issue is the following : how to quit a game with a physical button on GPIO. If you use the mk arcade joystick gpio, the gpio are mapped to a joystick. So you will have to be joystick 1 to use the hotkey + btn combo. Yes the hotkey combination is only available for joystick one. Plus, if you have this joystick declared, it could be detected and used for player 1 or 2, although you only want it to exit the game...
So it cannot work.

It could be useful to have an application that map gpio event to joystick keys. Here comes the ADAFRUIT RETROGAME that can do this job. So you just will have an other keyboard device, and you 'll have to change the exit_emulator_key in the retroarchcustom.cfg.

**Comment by guilegatan** [comment]
**+1** for "Adafruit-retrogame" support!
configuring keyboard-presses to specific GPIO-pins through the simple /boot/retrogame.txt file is VERY nice and would be a great add-on to RecalBox!
